<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mutation on Open Reading FramE annotation • MORFEE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Mutation on Open Reading FramE annotation">
<meta property="og:description" content="Annotation of mutation which generate new ATG
 From a VCF file annotated by ANNOVAR, MORFEE annotates mutation
 which generate new potential start codon (i.e. ATG).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MORFEE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.20200329</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/daissi/MORFEE/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="morfee" class="section level2">
<h2 class="hasAnchor">
<a href="#morfee" class="anchor"></a>MORFEE</h2>
<p><strong>MORFEE</strong> (<strong>Mutation on Open Reading FramE annotation</strong>) is a tool (R package) that, from a VCF file, detects and annotates single nucleotide variants creating premature ATG codons.</p>
<p>MORFEE algorithm is written in R language and can run on all operating systems which have an R interpreter (including Linux, macOS and Windows). MORFEE starts with a minimal VCF file (i.e. with at least the <em>chr</em>, <em>position</em>, <em>reference allele</em> and <em>alternate allele</em> fields) as an input, but can also work with already <a href="http://annovar.openbioinformatics.org/">ANNOVAR</a>-annotated (<a href="https://doi.org/10.1093/nar/gkq603">Wang et al., 2010</a>) VCF files. MORFEE has some R packages dependencies which are available on <a href="https://cran.r-project.org/">CRAN</a> or <a href="https://www.bioconductor.org/">Bioconductor</a> repositories and uses the <a href="https://www.gencodegenes.org/">GENCODE</a> database.</p>
<p>When using MORFEE, please cite our <a href="https://doi.org/10.1101/2020.03.29.012054">publication</a>.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="from-bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#from-bioconductor" class="anchor"></a>From Bioconductor</h3>
<p>Not yet available!</p>
</div>
<div id="from-github" class="section level3">
<h3 class="hasAnchor">
<a href="#from-github" class="anchor"></a>From GitHub</h3>
<pre><code><a href="https://devtools.r-lib.org//reference">library("devtools")
install_github("daissi/MORFEE")</a></code></pre>
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>In a first step, MORFEE reads the input VCF file and use ANNOVAR (that has then to be beforehand installed) through the wrapper function <em>vcf.annovar.annotation</em> to annotate all variants. This step is skipped if the input file has already been annotated. The minimal ANNOVAR annotations required by MORFEE are:</p>
<ul>
<li>
<strong>Func.refGene</strong> to extract 5’UTR variants.</li>
<li>
<strong>GeneDetail.refGene</strong> to extract transcript ID.</li>
</ul>
<div id="a--annotation-with-annovar-directly" class="section level3">
<h3 class="hasAnchor">
<a href="#a--annotation-with-annovar-directly" class="anchor"></a>A. Annotation with ANNOVAR directly</h3>
<pre><code># Annotation with ANNOVAR
table_annovar.pl \
        inst/extdata/VCF_example.vcf \
        ${PATH_ANNOVAR_DB} \
        -buildver hg19 \
        -remove -protocol refGene,gwasCatalog,avsnp150,clinvar_20190305,gnomad211_genome,dbnsfp35a \
        -operation gx,r,f,f,f,f -nastring . -vcfinput \
        -xref ${PATH_ANNOVAR_DB}/morfee.lof_metrics.by_gene.txt</code></pre>
</div>
<div id="b--annotation-with-annovar-through-morfee" class="section level3">
<h3 class="hasAnchor">
<a href="#b--annotation-with-annovar-through-morfee" class="anchor"></a>B. Annotation with ANNOVAR through MORFEE</h3>
<pre><code><a href="https://rdrr.io/pkg/VariantAnnotation/man">library(VariantAnnotation)
library(MORFEE)

my_raw_vcf &lt;- "inst/extdata/VCF_example.vcf"

# Annotation with ANNOVAR through the wrapper function
vcf.annovar.annotation(my_raw_vcf)</a></code></pre>
</div>
<div id="annotation-with-morfee" class="section level3">
<h3 class="hasAnchor">
<a href="#annotation-with-morfee" class="anchor"></a>Annotation with MORFEE</h3>
<pre><code><a href="https://rdrr.io/pkg/VariantAnnotation/man">library("VariantAnnotation")
library("MORFEE")

# Download database and create an R object used by MORFEE
MORFEE_DATA &lt;- get.morfee.data()

# To avoid too much Internet traffic, save locally this object
# save(MORFEE_DATA, file="MORFEE_DATA.RData")
# load("MORFEE_DATA.RData")

MY_VCF &lt;- "VCF_example.hg19_multianno.vcf"

# Read/Load the VCF into R
my_vcf &lt;- readVcf(MY_VCF)

# Perform the MORFEE annotation
my_vcf_morfee &lt;- morfee.annotation(my_vcf, MORFEE_DATA)

# Keep only variants which create a new upstream ATG sequence
my_vcf_morfee_NEW &lt;- my_vcf_morfee[!is.na(info(my_vcf_morfee)$MORFEE)]

# Write a VCF file with the new MORFEE field/annotation
writeVcf(my_vcf_morfee_NEW, filename="VCF_example.hg19_multianno.morfee_anno_NEW.vcf")

# Write an XLSX file with the new MORFEE field/annotation
vcf_2_xlsx(my_vcf_morfee_NEW, file="VCF_example.hg19_multianno.morfee_anno_NEW.xlsx")</a></code></pre>
<p>Please see below the description of the different columns of the xlsx file:</p>
<ul>
<li>
<strong>GeneDetail.refGene</strong>: Sequence Variant Nomenclature (ANNOVAR output). Please note, this nomenclature does not respect the <a href="https://varnomen.hgvs.org/">HGVS Recommendations</a> (<a href="https://doi.org/10.1002/humu.22981">doi:10.1002/humu.22981</a>). The position of the variant used with a prefix <em>c.</em> should be the position of the variant regarding the ATG position based one the coding sequence. In other words, the position on a sequence WITHOUT introns. ANNOVAR reports the position on the sequence WITH introns.</li>
<li>
<strong>orfSNVs</strong>: <em>in_frame</em> or <em>out_of_frame</em> to determine whether there is a frame shift between the new ATG and the reference one.</li>
<li>
<strong>Ratio_length_pred_obs</strong>: ratio between the predicted vs the observed length of the protein (&lt;1 for a shorter protein and &gt;1 for a longer protein). In case of several transcripts, only one value is given. That corresponding to the predicted length the closest to the observed length, i.e the ratio closest to 1.</li>
<li>
<strong>NewAALength</strong>: predicted length of the new protein in amino acids.</li>
<li>
<strong>MORFEE</strong>: MORFEE VCF-style annotation</li>
</ul>
<p><em>All fields below come from ANNOVAR annotation</em>:</p>
<ul>
<li>
<strong>pLI.refGene</strong>: <em>probability of Loss-of-function Intolerance</em> score from <a href="https://gnomad.broadinstitute.org/">gnomAD</a>
</li>
<li>
<strong>exp_lof.refGene</strong>: expected <em>predicted Loss-of-Function</em> variants from gnomAD</li>
<li>
<strong>oe_lof.refGene</strong>: <em>observed / expected</em> score</li>
<li>
<strong>oe_lof_lower.refGene</strong>: lower limit of 90% confidence interval of <em>oe</em> metric</li>
<li>
<strong>oe_lof_upper.refGene</strong>: upper limit of 90% confidence interval of <em>oe</em> metric</li>
<li>
<strong>gwasCatalog</strong>: gene-associated phenotype from <a href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a>
</li>
<li>
<strong>CLNDN</strong>: disease name from <a href="https://www.ncbi.nlm.nih.gov/variation/">ClinVar</a>
</li>
<li>
<strong>CLNDISDB</strong>: database name and identifier for the disease name from ClinVar</li>
<li>
<strong>CLNSIG</strong>: clinical significance from ClinVar</li>
<li>
<strong>AF</strong>: allele frequency from <a href="https://gnomad.broadinstitute.org/">gnomAD</a>
</li>
<li>
<strong>AF_popmax</strong>: maximum population allele frequency from gnomAD</li>
<li>
<strong>SIFT_pred</strong>: <em>Sorting Tolerant From Intolerant</em> function prediction from <a href="https://sites.google.com/site/jpopgen/dbNSFP">dbNSFP</a>
</li>
<li>
<strong>Polyphen2_HDIV_pred</strong>: Polyphen2 function prediction based on HumDiv from dbNSFP</li>
<li>
<strong>Polyphen2_HVAR_pred</strong>: Polyphen2 function prediction based on HumVar from dbNSFP</li>
<li>
<strong>LRT_pred</strong>: <em>Likelihood Ratio Test</em> function prediction from dbNSFP</li>
<li>
<strong>MutationTaster_pred</strong>: MutationTaster function prediction from dbNSFP</li>
<li>
<strong>MutationAssessor_pred</strong>: MutationAssessor function prediction from dbNSFP</li>
<li>
<strong>FATHMM_pred</strong>: <em>Functional Analysis through Hidden Markov Models</em> function prediction from dbNSFP</li>
<li>
<strong>PROVEAN_pred</strong>: <em>Protein Variation Effect Analyzer</em> function prediction from dbNSFP</li>
<li>
<strong>MetaSVM_pred</strong>: <em>Support Vector Machine</em> combination of function prediction score from dbNSFP</li>
<li>
<strong>MetaLR_pred</strong>: <em>Logistic Regression</em> combination of function prediction score from dbNSFP</li>
<li>
<strong>M.CAP_pred</strong>: <em>Mendelian Clinically Applicable Pathogenicity</em> likelihood score from dbNSFP</li>
<li>
<strong>REVEL_rankscore</strong>: <em>Rare Exome Variant Ensemble Learner</em> pathogenicity score from dbNSFP</li>
<li>
<strong>MutPred_rankscore</strong>: pathogenicity score from dbNSFP</li>
<li>
<strong>CADD_phred</strong>: <em>Combined Annotation Dependent Depletion</em> pathogenicity score from dbNSFP</li>
<li>
<strong>fathmm.MKL_coding_pred</strong>: fathmm-MKL function prediction based on HumDiv from dbNSFP</li>
<li>
<strong>integrated_fitCons_score</strong>: fitCons function prediction based on HumDiv from dbNSFP</li>
<li>
<strong>GERP.._RS</strong>: GERP++ RS conservation score from dbNSFP</li>
<li>
<strong>phyloP100way_vertebrate_rankscore</strong>: phyloP conservation score from dbNSFP</li>
<li>
<strong>phyloP20way_mammalian_rankscore</strong>: phyloP conservation score from dbNSFP</li>
<li>
<strong>phastCons100way_vertebrate_rankscore</strong>: phastCons conservation score from dbNSFP</li>
<li>
<strong>phastCons20way_mammalian_rankscore</strong>: conservation score from dbNSFP</li>
<li>
<strong>SiPhy_29way_logOdds</strong>: SiPhy conservation score from dbNSFP</li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/daissi/MORFEE/">https://​github.com/​daissi/​MORFEE/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/daissi/MORFEE/issues">https://​github.com/​daissi/​MORFEE/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;=3)</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing MORFEE</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Dylan Aïssi <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5858-6468" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Omar Soukarieh <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4923-4353" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>David-Alexandre Trégouët <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9084-7800" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/daissi/MORFEE"><img src="https://travis-ci.org/daissi/MORFEE.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://github.com/daissi/MORFEE/actions"><img src="https://github.com/daissi/MORFEE/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle: maturing"></a></li>
<li><a href="https://doi.org/10.1101/2020.03.29.012054"><img src="https://zenodo.org/badge/doi/10.1101/2020.03.29.012054.svg" alt="DOI"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Dylan Aïssi, Omar Soukarieh, David-Alexandre Trégouët.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
